pkg/controllers/proxyconfig/proxyconfig_controller.go

```diff
@@ -0,0 +1,134 @@
+}
+
+func (p *proxyConfigChecker) sync(ctx context.Context, _ factory.SyncContext) error {
+	if !p.isProxyConfigured() {
+		// If proxy is not configured, then it is a no-op.
+		return nil
+	}
+
+	route, err := p.routeLister.Routes("openshift-authentication").Get("oauth-openshift")
+	if err != nil {
+		return err
+	}
+
+	endpointURL, err := url.Parse(route.Spec.Host)
+	if err != nil {
+		return err
+	}
+
+	withProxySuccess := p.isEndpointReachable(ctx, "https://"+route.Spec.Host+"/healthz", p.proxyFunc)
+	withoutProxySuccess := p.isEndpointReachable(ctx, "https://"+route.Spec.Host+"/healthz", nil)
+	noProxyMatchesEndpoint := strings.Contains(p.proxyConfig.NoProxy, endpointURL.Hostname())
+
+	if noProxyMatchesEndpoint && !withoutProxySuccess {
+		if withProxySuccess {
+			return fmt.Errorf("endpoint(%q) found in NO_PROXY(%q) but can only be reached with proxy", endpointURL.Hostname(), p.proxyConfig.NoProxy)
+		}
+		return fmt.Errorf("endpoint(%q) found in NO_PROXY(%q) is not reachable with/without proxy", endpointURL.Hostname(), p.proxyConfig.NoProxy)
+	}
+
+	if !noProxyMatchesEndpoint && !withProxySuccess {
+		if withoutProxySuccess {
+			return fmt.Errorf("endpoint(%q) absent in NO_PROXY(%q) can only be reached without proxy", endpointURL.Hostname(), p.proxyConfig.NoProxy)
+		}
+		return fmt.Errorf("endpoint(%q) is not reachable with/without proxy", endpointURL.Hostname())
+	}
+
+	return nil
+}
+
+func (p *proxyConfigChecker) isEndpointReachable(ctx context.Context, endpointURL string, proxyFunc func(req *http.Request) (*url.URL, error)) bool {
+	req, err := http.NewRequest(http.MethodGet, endpointURL, nil)
+	if err != nil {
+		klog.V(4).Infof("failed to create new http request: %v", err)
+		return false
+	}
+	reqCtx, cancel := context.WithTimeout(ctx, 10*time.Second) // avoid waiting forever
+	defer cancel()
+	req.WithContext(reqCtx)
+
+	client := &http.Client{
+		Timeout: 5 * time.Second,
+		Transport: &http.Transport{
+			Proxy: proxyFunc,
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		},
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		klog.V(4).Infof("failed to make request to endpoint: %v", err)
+		return false
+	}
+
+	if resp.StatusCode > 299 || resp.StatusCode < 200 {
+		klog.V(4).Infof("%q returned %q", endpointURL, resp.Status)
+		return false
+	}
+	return true
+}
+
+func (p *proxyConfigChecker) isProxyConfigured() bool {
+	return len(p.proxyConfig.HTTPProxy) == 0 && len(p.proxyConfig.HTTPSProxy) == 0
+}
+
+func (p *proxyConfigChecker) proxyFunc(req *http.Request) (*url.URL, error) {
+	if req.URL.Scheme == "https" {
+		proxyURL, err := url.Parse(p.proxyConfig.HTTPSProxy)
+		if err == nil {
+			return proxyURL, nil
+		}
+		klog.V(4).Infof("failed to parse https proxy %q", p.proxyConfig.HTTPSProxy)
+	}
+
+	proxyURL, err := url.Parse(p.proxyConfig.HTTPProxy)
+	if err != nil {
+		return nil, err
+	}
+	return proxyURL, nil
+}
```

Comment:
or you don't cache the httproxy.FromEnvironment result and just use the function instead

Reply:
I don't think we can use `httpproxy.FromEnvironment` here. With endpoint matches NO_PROXY, then this function would not use the proxy. But for one specific case, this controller should reach endpoint via proxy even if endpoint is present in NO_PROXY. Does that make sense?

Reply:
I thought this was more about trying to connect to the endpoint with and without the proxy configuration. Why would we try ignore specifically configured NO_PROXY?

Reply:
Imagine the case where NO_PROXY is misconfigured with endpoint. In that case, connecting to an endpoint with proxy will help troubleshoot this issue. Otherwise, testing the endpoint without proxy will not be any different from testing with proxy and endpoint configured in NO_PROXY. Does that sound reasonable to you? It is possible that I misunderstood the requirement too.

Reply:
That makes sense